### PR TITLE
Add support for getting configuration from a secret

### LINKED ISSUE #13: Be able to source CONFIG_FILE_DATA from a secret (datastax/cass-operator#413)
Issue to track datastax/cass-operator#413

### PR DESCRIPTION
**Pull request by jsanda**
Fixes #13

This PR adds support for specifying Cassandra configuration with a secret. Here is an example to illustrate the usage. 

cassdc.yaml:

```yaml
apiVersion: cassandra.datastax.com/v1beta1
kind: CassandraDatacenter
metadata:
  name: dc1
spec:
  clusterName: test
  configSecret: cassdc-config
  size: 1
  storageConfig:
    cassandraDataVolumeClaimSpec:
      accessModes:
      - ReadWriteOnce
      resources:
        requests:
          storage: 5Gi
      storageClassName: standard
```

config-secret.yaml:

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: cassdc-config
type: Opaque
stringData:
  config: |-
    {
      "cassandra-yaml": {
        "roles_update_interval_in_ms": 65000,
        "read_request_timeout_in_ms": 5000
      },
      "jvm-options": {
        "initial_heap_size": "512M",
        "max_heap_size": "512M"
      }
    }
```

Here is a summary of how things work:

* If the secret does not exist reconciliation fails
* Fetch the secret specified by `configSecret`
* Add an annotation with the name of the datacenter to the secret
    * This annotation is used to set up a watch
* Get the config from the secret and apply additional properties 
    * Note that this is existing functionality but needs to be pointed out
* Get or create the _datacenter config secret_
    * This secret is created by cass-operator. It contains the combined config.
* Check to see if the config has changed
* If the config has changed
    * Add annotation with a hash of the datacenter config secret to the datacenter object
    * Patch the datacenter
    * Persist the datacenter config secret
* When the statefulset is constructed check to see if `configSecret` is set. It takes precedence over the inline config
    * Add the `CONFIG_FILE_DATA env var to the server config init container and the value comes from the datacenter config secret
    * Add a `CONFIG_HASH` env that contains the config has
        * This is to facilitate statefulset updates when the config secret is updated

### CODE REVIEW

File: operator/pkg/apis/cassandra/v1beta1/cassandradatacenter_types.go [hunk 1]
```diff
@@ -85,6 +85,8 @@ type CassandraDatacenterSpec struct {
 	// +kubebuilder:pruning:PreserveUnknownFields
 	Config json.RawMessage `json:"config,omitempty"`
 
+	ConfigSecret string `json:"configSecret,omitempty"`
```
Comment (burmanm) [inline operator/pkg/apis/cassandra/v1beta1/cassandradatacenter_types.go:111]: Missing description

File: operator/pkg/controller/cassandradatacenter/cassandradatacenter_controller.go [hunk 2]
```diff
@@ -120,6 +120,56 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
 		return err
 	}
 
+	dcAnnotation := "cassandra.datastax.com/datacenter"
```
Comment (burmanm) [inline operator/pkg/controller/cassandradatacenter/cassandradatacenter_controller.go]: This is used as a label (named DatacenterLabel from the cassdctypes) often, should probably use that one and explain why it's an annotation here instead of label.

File: operator/pkg/controller/cassandradatacenter/cassandradatacenter_controller.go [hunk 3]
```diff
@@ -120,6 +120,56 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
...
+	
+	configSecretMapFn := handler.ToRequestsFunc(func(mapObj handler.MapObject) []reconcile.Request {
+		log.Info("config secret watch called", "Secret", mapObj.Meta.GetName())
+
+		requests := make([]reconcile.Request, 0)
+		secret := mapObj.Object.(*corev1.Secret)
+		v, ok := secret.Annotations[dcAnnotation]
```
Comment (burmanm) [inline operator/pkg/controller/cassandradatacenter/cassandradatacenter_controller.go]: Merge with the next if

File: operator/pkg/reconciliation/construct_podtemplatespec.go [hunk 4]
```diff
@@ -325,6 +329,45 @@ func buildInitContainers(dc *api.CassandraDatacenter, rackName string, baseTempl
...
+
+		if configHash, ok := dc.Annotations[configHashAnnotation]; ok {
+			envVars = append(envVars, corev1.EnvVar{
+				Name: "CONFIG_HASH",
+				Value: configHash,
+			})
+		} else {
+			return nil, fmt.Errorf("datacenter %s is missing %s annotation", dc.Name, configHashAnnotation)
+		}
+		return envVars, nil
```
Comment (burmanm) [inline operator/pkg/reconciliation/construct_podtemplatespec.go]: Could be inside the if clause and that else removed.

File: operator/pkg/reconciliation/reconcile_configsecret.go [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+
+		rc.ReqLogger.Info("updating datacenter config secret", "ConfigSecret", dcConfigSecret.Name)
+		dcConfigSecret.Data["config"] = config
+
+		if exists {
+			if err := rc.Client.Update(rc.Ctx, dcConfigSecret); err != nil {
+				rc.ReqLogger.Error(err,"failed to update datacenter config secret", "ConfigSecret", dcConfigSecret.Name)
+				return result.Error(err)
+			}
+		} else {
```
Comment (burmanm) [inline operator/pkg/reconciliation/reconcile_configsecret.go]: Unnecessary else

File: operator/pkg/reconciliation/reconcile_configsecret.go [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+// checkDatacenterNameAnnotation Checks to see if the secret has the
+// cassandra.datastax.com/datacenter annotation with the datacenter name as its value. If
+// the secret does not have the annotation, it is added, and the secret is patched. The
+// secret should be the one specifiied by ConfigSecret.
+func (rc *ReconciliationContext) checkDatacenterNameAnnotation(secret *corev1.Secret) error {
+	annotation := "cassandra.datastax.com/datacenter"
+
+	if v, ok := secret.Annotations[annotation]; ok &&  v == rc.Datacenter.Name {
+		return nil
+	} else {
```
Comment (burmanm) [inline operator/pkg/reconciliation/reconcile_configsecret.go]: .. remove this one also.

File: operator/pkg/reconciliation/construct_podtemplatespec.go [hunk 7]
```diff
@@ -325,6 +329,45 @@ func buildInitContainers(dc *api.CassandraDatacenter, rackName string, baseTempl
...
+				SecretKeyRef: &corev1.SecretKeySelector{
+					LocalObjectReference: corev1.LocalObjectReference{
+						Name: getDatacenterConfigSecretName(dc),
+					},
+					Key: "config",
+				},
+			},
+		})
+
+		configHashAnnotation := "cassandra.datastax.com/config-hash"
```
[the remark this answers is not in the document]
Comment (jsanda) [reply operator/pkg/reconciliation/construct_podtemplatespec.go]: What do you mean?

File: .github/workflows/kindIntegTest.yml [hunk 8]
```diff
@@ -88,6 +88,8 @@ jobs:
         - additional_seeds
         - cluster_wide_install
         - config_change
+        - internode_encryption
```
Comment (burmanm) [inline .github/workflows/kindIntegTest.yml]: Remove, lets not re-enable it ;)

Comment (jsanda) [reply .github/workflows/kindIntegTest.yml]: Just my way of saying, are you sure.... 😆

File: operator/pkg/reconciliation/reconcile_configsecret.go [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+	patch := client.MergeFrom(secret.DeepCopy())
+	secret.Annotations[api.DatacenterAnnotation] = rc.Datacenter.Name
+	return rc.Client.Patch(rc.Ctx, secret, patch)
+}
+
+// updateConfigHashAnnotation Adds the config hash annotation to the datacenter. The value
+// of the annotation is a hash of the secret. The datacenter is then patched. Note that the
+// secret should be the datacenter secret created by the operator, NOT the secret specified
+// by ConfigSecret.
+func (rc *ReconciliationContext) updateConfigHashAnnotation(secret *corev1.Secret) error {
```
Comment (burmanm) [inline operator/pkg/reconciliation/reconcile_configsecret.go:97]: How do we ensure that the copy secret has the same security as the original secret? If the data is sensitive, we don't want this copy to be viewable anymore than the real secret. Or if the secret came from Hashicorp Vault or some other injection method (I don't know how the customer handles them if they're really sensitive).

File: tests/testdata/default-three-rack-three-node-dc.yaml [hunk 10]
```diff
@@ -17,18 +17,11 @@ spec:
         resources:
           requests:
             storage: 1Gi
-  resources:
```
Comment (burmanm) [inline tests/testdata/default-three-rack-three-node-dc.yaml:20]: Why are these removed?

File: operator/pkg/reconciliation/reconcile_configsecret.go (hunk 9 above)
Comment (jsanda) [reply operator/pkg/reconciliation/reconcile_configsecret.go:97]: >How do we ensure that the copy secret has the same security as the original secret? If the data is sensitive, we don't want this copy to be viewable anymore than the real secret. 

Great question. I think this falls more on the administrator or whoever is setting up RBAC for users and service accounts. The operator already has permissions to read/write secrets in the namespace in which it is deployed. I think handling other RBAC falls outside the scope of the operator.

File: tests/testdata/default-three-rack-three-node-dc.yaml (hunk 10 above)
Comment (jsanda) [reply tests/testdata/default-three-rack-three-node-dc.yaml:20]: I thought some of the initial failures may have been to do the limited resources. The max heap was basically the same as the memory limit specified under `resources`, leaving very little space for off-heap memory.

**Comment by burmanm** [comment]
Just last questions:

If the secret is created after the Cassdc, what happens? I guess it will not be picked up, since it's not owned by the Cassdc - but do we reconcile all the secrets if that annotation was applied to it?

And if the secret is deleted - what happens?

(it's fine if these are documentation only issues - but just wanted to know the intended behavior)

**Comment by jsanda** [comment]
>If the secret is created after the Cassdc, what happens?

I assume you are referring to the secret specified by the `ConfigSecret` property. Reconciliation fails. StatefulSets will not be created. 

>do we reconcile all the secrets if that annotation was applied to it?

Changes to any secret having the `cassandra.datastax.com/datacenter` annotation will trigger reconciliation on the CassandraDatacenter. The CassandraDatacenter is determined by the value of the annotation here:

```go
if v, ok := secret.Annotations[api.DatacenterAnnotation]; ok {
			log.Info("adding reconciliation request for config secret", "Secret", secret.Name)
			requests = append(requests, reconcile.Request{
				NamespacedName: types.NamespacedName{
					Namespace: secret.Namespace,
					Name: v,
				},
			})
		}
```

>And if the secret is deleted - what happens?

Same as above - reconciliation fails because `ConfigSecret` points to a secret that does not exist. Let's say you unset `ConfigSecret`. Then we fall back to using `Config` or defaults.
